fortuna/src/keeper.rs

```diff
@@ -0,0 +1,348 @@
+            StreamExt,
+        },
+        types::U256,
+    },
+    std::sync::Arc,
+    tokio::{
+        sync::mpsc,
+        time::{
+            self,
+            sleep,
+            Duration,
+        },
+    },
+    tracing,
+};
+
+pub async fn process_event(
+    event: RequestedWithCallbackEvent,
+    chain_config: &api::BlockchainState,
+    contract: &Arc<SignablePythContract>,
+    nonce_manager: &Arc<NonceManagerMiddleware<Provider<Http>>>,
+    gas_limit: U256,
+) -> Result<()> {
+    let provider_revelation = chain_config.state.reveal(event.sequence_number)?;
+
+    let sim_res = chain_config
+        .contract
+        .simulate_reveal(
+            event.provider_address,
+            event.sequence_number,
+            event.user_random_number,
+            provider_revelation,
+        )
+        .await;
+
+    match sim_res {
+        Ok(_) => {
+            let res = contract
+                .reveal_with_callback_wrapper(
+                    event.provider_address,
+                    event.sequence_number,
+                    event.user_random_number,
+                    provider_revelation,
+                    nonce_manager.next(),
+                    gas_limit,
+                )
+                .await;
+            match res {
+                Ok(_) => {
+                    tracing::info!(
+                        "Revealed for provider: {} and sequence number: {} with res: {:?}",
+                        event.provider_address,
+                        event.sequence_number,
+                        res
+                    );
+                }
+                Err(e) => {
+                    // TODO: find a way to handle different errors.
+                    tracing::error!(
+                        "Error while revealing for provider: {} and sequence number: {} with error: {:?}",
+                        event.provider_address,
+                        event.sequence_number,
+                        e
+                    );
+                }
+            }
+        }
+        Err(e) => {
+            tracing::error!(
+                "Error while simulating reveal for provider: {} and sequence number: {} \n res: {:?}",
+                event.provider_address,
+                event.sequence_number,
+                e
+            );
+        }
+    }
+    Ok(())
+}
+
+pub async fn handle_backlog(
+    chain_id: String,
+    latest_safe_block: u64,
+    nonce_manager: Arc<NonceManagerMiddleware<Provider<Http>>>,
+    contract: Arc<SignablePythContract>,
+    gas_limit: U256,
+    chain_config: api::BlockchainState,
+) -> Result<()> {
+    tracing::info!("Starting backlog handler for chain: {}", &chain_id);
+    let backlog_blocks: u64 = 10_000;
+    let blocks_at_a_time = 100;
+
+    let mut from_block = if backlog_blocks > latest_safe_block {
+        0
+    } else {
+        latest_safe_block - backlog_blocks
+    };
+    let last_block = latest_safe_block;
+
+    while from_block < last_block {
```

Comment:
shouldn't this be `from_block <= last_block`?

Reply:
good catch. thanks!